### PR TITLE
I1863 idklu casadi

### LINKED ISSUE #1863: idklu solver for `convert_to_casadi=True`
**Summary**

The current idklu solver goes back to python whenever it needs to calculate the rhs, algebraic or event equations. It also runs very slow when using `convert_to_casadi=True`. This issue aims to develop a new idklu solver that directly calls casadi functions from C++, with ideally no memory copying to/from casadi arrays

 **Motivation**

We want to get the superior event handling capabilities of the ida solvers with efficient casadi evaluation of equations

### LINKED ISSUE #2008: [Bug]: expected size of InputParameter set incorrectly in `pybamm.Discretisation._process_symbol`
### PyBaMM Version

22.3

### Python Version

3.8

### Describe the bug

In `pybamm.Discretisation._process_symbol`, the `pybamm.InputParameter` is processed in lines 1038-1045, like so:

```python
elif isinstance(symbol, pybamm.InputParameter):
            # Return a new copy of the input parameter, but set the expected size
            # according to the domain of the input parameter
            expected_size = self._get_variable_size(symbol)
            new_input_parameter = pybamm.InputParameter(
                symbol.name, symbol.domain, expected_size
            )
            return new_input_parameter
```

This will result in an incorrect `expected_size` if the size of the InputParameter is set using `expected_size` argument, rather than a domain.

### Steps to Reproduce

```python
model = pybamm.BaseModel()
model.convert_to_format = form
u1 = pybamm.Variable("u1")
u2 = pybamm.Variable("u2")
u3 = pybamm.Variable("u3")
v = pybamm.Variable("v")
a = pybamm.InputParameter("a")
b = pybamm.InputParameter("b", expected_size=2)
model.rhs = {u1: a * v, u2: pybamm.Index(b, 0), u3: pybamm.Index(b, 1)}
model.algebraic = {v: 1 - v}
model.initial_conditions = {u1: 0, u2: 0, u3: 0, v: 1}

disc = pybamm.Discretisation()
disc.process_model(model)
```

fails with error

### Relevant log output

```shell
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/martinjrobins/git/PyBaMM/pybamm/discretisations/discretisation.py", line 821, in process_symbol
    return self._discretised_symbols[symbol.id]
KeyError: -9213461660596305478

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/martinjrobins/git/PyBaMM/pybamm/discretisations/discretisation.py", line 821, in process_symbol
    return self._discretised_symbols[symbol.id]
KeyError: -8616933861355187750

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/martinjrobins/git/PyBaMM/pybamm/expression_tree/symbol.py", line 770, in evaluate_for_shape
    return self._saved_evaluate_for_shape
AttributeError: 'Index' object has no attribute '_saved_evaluate_for_shape'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "tests/unit/test_solvers/test_idaklu_solver.py", line 162, in test_input_params
    disc.process_model(model)
  File "/Users/martinjrobins/git/PyBaMM/pybamm/discretisations/discretisation.py", line 208, in process_model
    rhs, concat_rhs, alg, concat_alg = self.process_rhs_and_algebraic(model)
  File "/Users/martinjrobins/git/PyBaMM/pybamm/discretisations/discretisation.py", line 611, in process_rhs_and_algebraic
    processed_rhs = self.process_dict(model.rhs)
  File "/Users/martinjrobins/git/PyBaMM/pybamm/discretisations/discretisation.py", line 799, in process_dict
    processed_eqn = self.process_symbol(eqn)
  File "/Users/martinjrobins/git/PyBaMM/pybamm/discretisations/discretisation.py", line 823, in process_symbol
    discretised_symbol = self._process_symbol(symbol)
  File "/Users/martinjrobins/git/PyBaMM/pybamm/discretisations/discretisation.py", line 886, in _process_symbol
    disc_child = self.process_symbol(child)
  File "/Users/martinjrobins/git/PyBaMM/pybamm/discretisations/discretisation.py", line 823, in process_symbol
    discretised_symbol = self._process_symbol(symbol)
  File "/Users/martinjrobins/git/PyBaMM/pybamm/discretisations/discretisation.py", line 973, in _process_symbol
    return symbol._unary_new_copy(disc_child)
  File "/Users/martinjrobins/git/PyBaMM/pybamm/expression_tree/unary_operators.py", line 332, in _unary_new_copy
    new_index = self.__class__(child, self.index, check_size=False)
  File "/Users/martinjrobins/git/PyBaMM/pybamm/expression_tree/unary_operators.py", line 294, in __init__
    super().__init__(name, child)
  File "/Users/martinjrobins/git/PyBaMM/pybamm/expression_tree/unary_operators.py", line 36, in __init__
    super().__init__(name, children=[child], domains=domains)
  File "/Users/martinjrobins/git/PyBaMM/pybamm/expression_tree/symbol.py", line 229, in __init__
    self.test_shape()
  File "/Users/martinjrobins/git/PyBaMM/pybamm/expression_tree/symbol.py", line 991, in test_shape
    self.shape_for_testing
  File "/Users/martinjrobins/git/PyBaMM/pybamm/expression_tree/symbol.py", line 967, in shape_for_testing
    evaluated_self = self.evaluate_for_shape()
  File "/Users/martinjrobins/git/PyBaMM/pybamm/expression_tree/symbol.py", line 772, in evaluate_for_shape
    self._saved_evaluate_for_shape = self._evaluate_for_shape()
  File "/Users/martinjrobins/git/PyBaMM/pybamm/expression_tree/unary_operators.py", line 339, in _evaluate_for_shape
    return self._unary_evaluate(self.children[0].evaluate_for_shape())
  File "/Users/martinjrobins/git/PyBaMM/pybamm/expression_tree/unary_operators.py", line 328, in _unary_evaluate
    return child[self.slice]
TypeError: 'float' object is not subscriptable

----------------------------------------------------------------------
```
```

### PR DESCRIPTION
**Pull request by martinjrobins**
# Description

Adds a casadi version of the idklu solver, currently getting approx 2 x speed-up over current casadi solver with fast events, but this still needs to be properly benchmarked

As part of the above, the `process` function in the base solver now also returns a function that evaluates the jacobian times a vector (e.g. `casadi.jtimes`). This was neccessary to implement the sensitivity equations that IDA expects, but will also be useful when we add the option of using iterative solvers.

update: also fixed a discretisation bug that occurs when an input parameter is indexed, see #2008

Fixes #1863 
Fixes #2008 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] New feature (non-breaking change which adds functionality)

# Key checklist:

- [x] No style issues: `$ flake8`
- [x] All tests pass: `$ python run-tests.py --unit`
- [x] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: tests/unit/test_solvers/test_idaklu_solver.py [hunk 1]
```diff
@@ -46,13 +53,148 @@ def test_ida_roberts_klu(self):
             true_solution = 0.1 * solution.t
             np.testing.assert_array_almost_equal(solution.y[0, :], true_solution)
 
-    @unittest.skipIf(not pybamm.have_jax(), "jax or jaxlib is not installed")
+    def test_model_events(self):
+        for form in ["python", "casadi", "jax"]:
+            if form == "jax" and not pybamm.have_jax():
+                continue
```
Comment (tinosulzer) [inline tests/unit/test_solvers/test_idaklu_solver.py:59]: shouldn't this line not be caught in coverage?

**Review by tinosulzer — APPROVED** [review]
Thanks @martinjrobins ! I didn't look too closely at the C++ code (since I can't read it), but the tests look good

File: tests/unit/test_solvers/test_idaklu_solver.py (hunk 1 above)
Comment (martinjrobins) [reply tests/unit/test_solvers/test_idaklu_solver.py:59]: when we measure the coverage it won't get to the continue, but its needed for the other test suites, for which jax might not be available.

Comment (martinjrobins) [reply tests/unit/test_solvers/test_idaklu_solver.py:59]: well, its in the test, I assume its not necessary to cover the actual test code :) Who watches the watchers and all that....

**Pull request merged**